TinyAlsa: Add support for testing using tinycap and tinyplay.

gbarkadiusz opened the pull request:
Added option `SOF_ALSA_TOOL` to specify whether to use the `TinyASLA` or `ALSA` testing tool. For now, it is only for capturing and playback.
The default tool is ALSA.

This PR introduces also a generic wrapper  `tinyalsa-wrapper.sh` to define `SOF_ALSA_TOOL` in environments where setting environment variables directly is not possible. This is useful for frameworks or systems with limited configuration capabilities.

golowanow commented on test-case/check-playback.sh:
```diff
@@ -91,12 +100,20 @@ do
+                       die "aplay on PCM $dev failed at $i/$loop_cnt."
+                    }
+                else
+                    echo "Unknown tool"
```
do this check for correct $tool earlier outside of the loop (e.g. L74)

marc-hb reviewed (commented):
This is not scalable at all. It's adding support in only 2 tests and there's already a lot of copy/paste/diverge.

The non-duplicated way is to make `aplay_opts` into a smarter function that implements all the logic in only one place.

redzynix commented on case-lib/lib.sh:
```diff
@@ -754,15 +754,43 @@ func_lib_check_pa()
+    if $use_tinyalsa; then
+        card_nr=$(echo "$dev" | cut -d ':' -f2 | cut -d ',' -f1)
+        dev_nr=$(echo "$dev" | cut -d ',' -f2)
+        sox -n -r "$rate" -c "$channel" noise.wav synth "$duration" white
```
IMHO I would declare channel, rate, duration, and other alsa related variables in lib.sh to avoid reference before the assignment.
https://github.com/thesofproject/sof-test/blob/main/test-case/check-playback.sh#L78C9-L84C52

gbarkadiusz replied:
Please take a look.

marc-hb commented:
Avoid "spamming" the PR with force-pushes as this brings down limited hardware resources.

golowanow commented on test-case/check-capture.sh:
```diff
@@ -56,10 +59,16 @@ duration=${OPT_VAL['d']}
 start_test
 logger_disabled || func_lib_start_log_collect
 
+if [[ "$tool" != "tinyalsa" ]] && [[ "$tool" != "alsa" ]]; then
```
please consider just to add an optional environment variable (e.g. `SOF_ALSA_TOOL`) and make all the choices, checks, etc. at lib.sh only, so changing nothing in test cases.

marc-hb replied:
Agreed, this looks like it should be a global setting of some kind. I don't see why a user would need to keep jumping back and forth between tinyalsa and regular depending on which test they run. And if they really do, they could still prefix and override each run with the global variable:

```shell
./test-case/check-something.sh
SOF_TINY_ALSA=true ./test-case/check-something-else.sh
./test-case/check-another-thing.sh

```

golowanow commented:
> Avoid "spamming" the PR with force-pushes as this brings down limited hardware resources.

also, shellcheck can be run locally, before push to PR.

marc-hb commented:
Don't forget the `-x` option, as in: `shellcheck -x somefile.sh`, otherwise you get a lot of false positives. I wonder why `-x` is not ON by default.

If you forget which option it is, just look in either the CI logs or in `git grep shellcheck -- .github/` and run the same. True not just with shellcheck. That's one big advantage of GitHub actions over Jenkins and similar: no "CI secret".

(there can also be drawbacks having everything in the same git repo, I digress)

gbarkadiusz commented:
@marc-hb @golowanow I've changed the approach according to your suggestions. Now, only the `lib.sh` has been changed. To force the usage of `tinyalsa` as the testing tool, the testcase should be run with `SOF_TINY_ALSA=true` before the testcase.
e.g. `SOF_TINY_ALSA=true ./test-case/check-playback.sh`
Moreover, I've added the declaration of ALSA-related variables to `lib.sh` to avoid reference before assignment, as @redzynix mentioned.

marc-hb commented on case-lib/lib.sh:
```diff
@@ -752,17 +752,37 @@ func_lib_check_pa()
+rate=$(func_pipeline_parse_value "$idx" rate)
+fmt_elem=$(func_pipeline_parse_value "$idx" fmt_elem)
+dev=$(func_pipeline_parse_value "$idx" dev)
+duration=$(func_pipeline_parse_value "$idx" duration)
```
There are 2 problems here that can be fortunately and easily solved with a single solution: moving these to a new "init" function explicitly invoked by each test that needs these. Whatever the question, a function is always the answer ;-)

1. These are run the moment `lib.sh` is `source`-d.
   1a. Individual tests have no control over when this is run
   1b. It is normally possible to `source case-lib/lib.sh` _from the command-line_ for interactive testing. To keep this functional, sourcing should run as little code as possible and have as few side effects as possible.

2. This code duplicates and clashes with what a lot of tests _already do_ with potential conflicts very difficult to debug. Just try `git grep func_pipeline_parse_value` and see.
 
Also, this addition is currently splitting `aplay_opts()` away from its comment/ "documentation".

By placing these in a new init function, you solve 1. and you can cleanup 2. for each test one by one. No need to do all of them at once, you can start with the few tests using `aplay_opts()` already.

marc-hb replied:
> These are run the moment lib.sh is source-d.

A similar problem was fixed in 7c2ede03c9be1b99, take a look at #1187 for context.

marc-hb commented on case-lib/lib.sh:
```diff
@@ -752,17 +752,37 @@ func_lib_check_pa()
-    arecord $SOF_ALSA_OPTS $SOF_ARECORD_OPTS "$@"
+    if [[ $SOF_TINY_ALSA == true ]]; then
+        card_nr=$(echo "$dev" | cut -d ':' -f2 | cut -d ',' -f1)
+        dev_nr=$(echo "$dev" | cut -d ',' -f2)
```
This is a copy of the code in `aplay_opts()`, move this to some new `set_tiny_alsa_params()` function.

marc-hb commented on case-lib/lib.sh:
```diff
@@ -752,17 +752,37 @@ func_lib_check_pa()
-    aplay $SOF_ALSA_OPTS $SOF_APLAY_OPTS "$@"
+    if [[ $SOF_TINY_ALSA == true ]]; then
+        card_nr=$(echo "$dev" | cut -d ':' -f2 | cut -d ',' -f1)
+        dev_nr=$(echo "$dev" | cut -d ',' -f2)
```
When parsing like this, it really helps the reader to add an example in a one-line comment:
```shell
# example: dev=hw:0,3
```

Also, prefer `printf %s` to `echo`:

```suggestion
        dev_nr=$(printf '%s' "$dev" | cut -d ',' -f2)
```
`echo` has a lot of pitfalls. It should not matter in this case but it's good practice. Plus everyone knows `printf` from other languages.

https://unix.stackexchange.com/questions/65803/why-is-printf-better-than-echo
https://www.shellcheck.net/wiki/SC2028
etc.

golowanow commented on case-lib/lib.sh:
```diff
@@ -752,17 +752,37 @@ func_lib_check_pa()
-    dlogc "aplay $SOF_ALSA_OPTS $SOF_APLAY_OPTS $*"
-    # shellcheck disable=SC2086
-    aplay $SOF_ALSA_OPTS $SOF_APLAY_OPTS "$@"
+    if [[ $SOF_TINY_ALSA == true ]]; then
```
just few additional suggestions from my side:
1.
```suggestion
    if [[ "$SOF_TINY_ALSA" = true ]]; then
```

2. `SOF_TINY_ALSA` - I'm not sure it is "scalable" enough as a bool variable in comparison to `SOF_ALSA_TOOL=tinyalsa` or whatever other/proprietary tool, a custom script tool etc. it might be in the future. 

3. anyway, document it: 
* [in comments](https://github.com/thesofproject/sof-test/blob/main/case-lib/lib.sh#L727-L753)
* [in README.md](https://github.com/thesofproject/sof-test/blob/main/README.md?plain=1#L56)

4. add `dlogc` similar to what the alsa branch does

gbarkadiusz replied:
Regarding point number 2, I think that it will always force the use of the SOF_ALSA_TOOL parameter, e.g., SOF_ALSA_TOOL=alsa.

marc-hb replied:
> SOF_ALSA_TOOL=alsa, which could break backward compatibility.

This can be a default value. Default values are often implemented like this:

```shell
: ${SOF_ALSA_TOOL:=alsa}
```

https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html
https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_06_02

Word expansion -> parameter expansion -> default value.

marc-hb commented on case-lib/lib.sh:
```diff
@@ -752,17 +752,84 @@ func_lib_check_pa()
+# Function to extract the card number and device number from $dev option (e.g., hw:0,10)
+parse_audio_device() {
+    # Extract the card number (e.g., "0" from hw:0,10)
+    card_nr=$(printf '%s' "$dev" | cut -d ':' -f2 | cut -d ',' -f1)
```
Avoiding global variables is sometimes difficult in shell but not in this case. In this case there's no reason to use a global variable.

```suggestion
    card_nr=$(printf '%s' "$1" | cut -d ':' -f2 | cut -d ',' -f1)
```

marc-hb commented on case-lib/lib.sh:
```diff
@@ -752,17 +752,84 @@ func_lib_check_pa()
+extract_format_number() {
+    # (e.g., extracting '16' from 'S16_LE')
+    # shellcheck disable=SC2154
+    format=$(printf '%s' "$fmt_elem" | grep '[0-9]\+' -o)
```
Same as above, and now you can drop the `shellcheck disable=SC2154`.
```suggestion
    format=$(printf '%s' "$1" | grep '[0-9]\+' -o)
```

marc-hb commented on case-lib/lib.sh:
```diff
@@ -752,17 +752,84 @@ func_lib_check_pa()
+
+# Initialize the parameters using for audio testing.
+initialize_audio_params()
+{
```
Same as above
```suggestion
{
   local idx="$1"
   
```
This will temporarily shadow any global `idx` as expected.

marc-hb commented on test-case/check-capture.sh:
```diff
@@ -68,19 +68,14 @@ for round in $(seq 1 $round_cnt)
-        type=$(func_pipeline_parse_value "$idx" type)
-        snd=$(func_pipeline_parse_value "$idx" snd)
+
+	initialize_audio_params
```
```suggestion
	initialize_audio_params "$idx"
```

marc-hb commented on test-case/check-playback.sh:
```diff
@@ -75,13 +75,8 @@ for round in $(seq 1 $round_cnt)
-        type=$(func_pipeline_parse_value "$idx" type)
-        snd=$(func_pipeline_parse_value "$idx" snd)
+
+        initialize_audio_params
```
```suggestion
        initialize_audio_params "$idx"
```

marc-hb commented on case-lib/lib.sh:
```diff
@@ -752,17 +752,84 @@ func_lib_check_pa()
+   pcm=$(func_pipeline_parse_value "$idx" pcm)
+   # shellcheck disable=SC2034
+   type=$(func_pipeline_parse_value "$idx" type)
+   # shellcheck disable=SC2034
```
I don't get this warning, are you sure?

If you are sure:
- add it only once right before the function and it will apply to whole function. No need to duplicate.
- Explain very briefly why it's disabled briefly before it. No one knows what SC2034 is by heart.

gbarkadiusz replied:
SC2034 (warning): pcm appears unused. Verify use (or export if used externally).

marc-hb commented on case-lib/lib.sh:
```diff
@@ -752,17 +752,84 @@ func_lib_check_pa()
-    aplay $SOF_ALSA_OPTS $SOF_APLAY_OPTS "$@"
+    if [[ "$SOF_ALSA_TOOL" = "tinyalsa" ]]; then
+        dlogc "tinyplay $*"
+	# shellcheck disable=SC2154
```
I don't see this warning. Did you use the `-x` option? `shellcheck -x ...`

marc-hb replied:
I think I know what's happening! Cause it just happened to me in a totally unrelated context. You must run `shellcheck -x case-lib/lib.sh` IN THE TOP-LEVEL sof-test directory! That will hopefully silence a few warnings.

marc-hb reviewed (commented):
Just a few minor issues left, otherwise looks good to me. Thanks for your persistence!

golowanow commented on test-case/check-capture-tinyalsa.sh:
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
```
1) please add some comments here to explain the purpose of this file as a wrapper to execute a test case script using  tinyalsa. As its parameters it expects a test case script file name (without path) in the first parameter, followed by other parameters for that test.

2) also, add here some simple preconditions to check the tinyalsa executables are present on the DUT.

golowanow commented on test-case/check-playback-tinyalsa.sh:
not needed with `check-tinyalsa.sh` wrapper

gbarkadiusz commented:
@golowanow I've added the changes for the volume-basic test. This test doesn't use lib.sh for executing the amixer command, so I've added if statements to make this test ready for testing mixers under TinyALSA. Please take a look and share your feedback.

marc-hb commented on test-case/volume-basic-test.sh:
```diff
@@ -36,13 +36,19 @@ func_opt_parse_option "$@"
+    if [[ "$SOF_ALSA_TOOL" = "alsa" ]]; then
+        pkill -9 aplay
+    elif [[ "$SOF_ALSA_TOOL" = "tinyalsa" ]]; then
+	pkill -9 tinyplay
```
indent

marc-hb commented on test-case/volume-basic-test.sh:
```diff
@@ -36,13 +36,19 @@ func_opt_parse_option "$@"
+    elif [[ "$SOF_ALSA_TOOL" = "tinyalsa" ]]; then
+	pkill -9 tinyplay
+    else
+	echo "Unknowns asla tool $SOF_ALSA_TOOL"
```
indent + typo

This fallback does not seem necessary. This is a clean up function, if SOF_ALSA_TOOL is wrong then there must have been another error message already printed much earlier than this.

marc-hb commented on test-case/volume-basic-test.sh:
```diff
@@ -56,20 +62,45 @@ rate=$(func_pipeline_parse_value 0 rate)
-dlogc "aplay -D $dev -c $channel -r $rate -f $fmt /dev/zero &"
-# play into background, this will wake up DSP and IPC. Need to clean after the test
-aplay -D "$dev" -c "$channel" -r "$rate" -f "$fmt" /dev/zero &
+if [[ "$SOF_ALSA_TOOL" = "alsa" ]]; then
```
I don't understand: you did a great work making `aplay_opts()` generic, why not use it here?

golowanow replied:
I have the same question

marc-hb commented on test-case/volume-basic-test.sh:
```diff
@@ -83,10 +114,17 @@ do
         for vol in "${volume_array[@]}"; do
-            dlogc "amixer -c$sofcard cset name='$volctrl' $vol"
-            amixer -c"$sofcard" cset name="$volctrl" "$vol" > /dev/null ||
+            if [[ "$SOF_ALSA_TOOL" = "alsa" ]]; then
```
Same question: why not use the new, upgraded and generic reset_volume() here?

marc-hb commented on test-case/volume-basic-test.sh:
```diff
@@ -96,8 +134,15 @@ do
 
-#clean up background aplay
-pkill -9 aplay || true
+if [[ "$SOF_ALSA_TOOL" = "alsa" ]]; then
```
Extract the killing part of `func_error_exit()` above into a new function and invoke it here.

marc-hb reviewed (commented):
> I've added the changes for the volume-basic test. This test doesn't use lib.sh for executing the amixer command, so I've added if statements ...

Why? You've done a great work upgrading aplay_opts(), making it generic, but at the last minute just before this PR was about to be all approved and merged, you add the same feature in volume-basic-test.sh while NOT using your brand new and shiny aplay_opts()? Strange?

Maybe there is a reason not to use aplay_opts() in volume-basic-test.sh. Maybe there is none. Either way, don't change volume-basic-test.sh in this PR and get this PR over with. Discuss volume-basic-test.sh in a new, separate PR.

marc-hb commented on test-case/check-tinyalsa.sh:
```diff
@@ -0,0 +1,34 @@
+# Preconditions: Check if tinyalsa executables are present on the DUT
+check_command "sox"
+check_command "tinycap"
+check_command "tinyplay"
```
I'm not convinced this wrapper is worth the added lines of code. If one of these commands are missing, `aplay_opts()` or equivalent should already fail with a decent "not found" error message. Is it not already the case? It should be.

Either way this is not strictly required, independent and can be moved to a future PR. This one has already lasted for a very long time :-)

marc-hb commented:
I suspect some are older than this PR but most of the few warnings left in https://github.com/thesofproject/sof-test/actions/runs/14243977340/job/39920223082?pr=1260 seem very easy to fix.

golowanow commented on README.md:
```diff
@@ -53,11 +53,13 @@ Usage: ./check-playback.sh [OPTION]
 and SOF_ARECORD_OPTS are passed to all aplay and arecord
 invocations. SOF_ALSA_OPTS parameters are passed to both aplay and
-arecord. Warning these environments variables do NOT support parameters
+arecord. SOF_ALSA_TOOL is used to select tool for testing.
```
TinyAlsa should also use SOF_(ALSA|APLAY|ARECORD)_OPTS assuming that they follow SOF_ALSA_TOOL value.
Please document this assumption here as well.

gbarkadiusz replied:
Done.

golowanow commented on case-lib/lib.sh:
```diff
@@ -752,17 +752,80 @@ func_lib_check_pa()
+    dev_nr=$(printf '%s' "$1" | cut -d ',' -f2)
+}
+
+# Function to extract the numeric format value from the $fmt_elem option
```
not clear what is `$fmt_elem` here ?

golowanow commented on case-lib/lib.sh:
```diff
@@ -752,17 +752,80 @@ func_lib_check_pa()
+# So far, supported tools are 'alsa' and 'tinyalsa'
+# To select appropriate tool, set SOF_ALSA_TOOL to one of above
+# before using 'aplay_opts' or 'arecord_opts' function.
+# (e.g., SOF_ALSA_TOOL=alsa)
```
```suggestion
Default is `SOF_ALSA_TOOL=`alsa`
```

golowanow commented on test-case/check-tinyalsa.sh:
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
+# This script serves as a wrapper to execute a test case script using tinyalsa.
```
please follow the same comment style here as in other test case scripts.

marc-hb replied:
The LICENSE is especially important. But: separate PR.

golowanow reviewed (commented):
some general asks from my side:
1) better to introduce TinyAlsa in one (this) PR to make all these changes consistent.
2) split changes into logically ordered commits.

gbarkadiusz commented:
@marc-hb restored to the previous version. I'll add `tinymix` in the separate pr.

gbarkadiusz commented:
> I suspect some are older than this PR but most of the few warnings left in https://github.com/thesofproject/sof-test/actions/runs/14243977340/job/39920223082?pr=1260 seem very easy to fix.

These warnings don't appear locally.
`shellcheck -x case-lib/lib.sh`

marc-hb commented:
> These warnings don't appear locally.
> shellcheck -x case-lib/lib.sh

Most of them are in check-playback.sh and check-capture.sh.

There's one in case-lib/lib.sh though and it's most basic type of warning so it's strange you don't see it.

Anyway they are all just missing quotes, super easy to fix. Just add double quotes! shellcheck even prints the correct code.

https://mywiki.wooledge.org/Quotes?highlight=%28quote%29#I.27m_Too_Lazy_to_Read.2C_Just_Tell_Me_What_to_Do

marc-hb commented:
> There's one in case-lib/lib.sh though and it's most basic type of warning so it's strange you don't see it.

I may have solved this mystery. I don't see these missing quotes with shellcheck 0.8 (Ubuntu 22.04) but I do see them with shellcheck 0.9 (Ubuntu 24.04). So, shellcheck bug fixes. The Ubuntu runner was recently updated (94bbe9d50b41)

What shellcheck -V ersion do you have?

gbarkadiusz commented:
@marc-hb Could you take a look, please.

marc-hb commented on case-lib/lib.sh:
```diff
@@ -1063,15 +1129,31 @@ set_alsa_settings()
+            fi
+        done
+    else
+         echo "Unknowns alsa tool $SOF_ALSA_TOOL"
```
Typo. Does not matter.

gbarkadiusz replied:
Has changed.

marc-hb reviewed (approved):
As I wrote before, I'm not convinced the wrapper adds much value but AFAICT it's optional. Hope it stays like this :-)

I have NOT looked at the test results (which are all red...) Obviously, this should not be merged if it introduces any regression.

Everything else looks good.

golowanow commented:
I'm looking through it again with recent probe run results.

golowanow commented on README.md:
```diff
@@ -53,11 +53,15 @@ Usage: ./check-playback.sh [OPTION]
-and SOF_ARECORD_OPTS are passed to all aplay and arecord
-invocations. SOF_ALSA_OPTS parameters are passed to both aplay and
-arecord. Warning these environments variables do NOT support parameters
+Some tests support the use of environment variables:
+SOF_ALSA_TOOL, SOF_ALSA_OPTS, SOF_APLAY_OPTS and SOF_ARECORD_OPTS (work in progress)
```
```suggestion
Some tests support these environment variables (work in progress):
```

golowanow commented on README.md:
```diff
@@ -53,11 +53,15 @@ Usage: ./check-playback.sh [OPTION]
+Some tests support the use of environment variables:
+SOF_ALSA_TOOL, SOF_ALSA_OPTS, SOF_APLAY_OPTS and SOF_ARECORD_OPTS (work in progress)
+  - SOF_ALSA_TOOL is used to select the audio tool for testing.
+  Set this variable to 'alsa' or 'tinyalsa' to choose between the ALSA and TinyALSA toolsets.
```
```suggestion
  Set this variable to 'alsa' (default value) or 'tinyalsa' to choose between the ALSA and TinyALSA toolsets.
```

golowanow commented on README.md:
```diff
@@ -53,11 +53,15 @@ Usage: ./check-playback.sh [OPTION]
+  Set this variable to 'alsa' or 'tinyalsa' to choose between the ALSA and TinyALSA toolsets.
+  - SOF_ALSA_OPTS contains optional parameters passed to both aplay and arecord.
+  - SOF_APLAY_OPTS and SOF_ARECORD_OPTS contain optional parameters passed specifically to aplay and arecord.
+These options aply to the selected tool (alsa or tinyalsa) based on the value of SOF_ALSA_TOOL 
```
```suggestion
These options are applied to the selected tool (alsa or tinyalsa) based on the value of SOF_ALSA_TOOL 
```

golowanow reviewed (changes requested):
@gbarkadiusz - why is shellcheck (binary!) included with its .txt files in this PR ?!
I'll continue review for the rest of changes, but -1 for now.

marc-hb commented:
>  why is shellcheck (binary!) included with its .txt files in this PR ?!

At first I felt ashamed I missed that and approved, then I realized this just came in the most recent force-push; _after_ I approved.

The reason is probably this above:
- https://github.com/thesofproject/sof-test/pull/1260#issuecomment-2786258632

@gbarkadiusz you should simply upgrade the entire distribution in your WSL container; not just shellcheck.

https://repology.org/project/shellcheck/versions

golowanow commented on test-case/check-tinyalsa.sh:
I think `check-tinyalsa.sh` should be moved to another commit - "Add support for TinyALSA"

marc-hb replied:
> I think check-tinyalsa.sh should be moved to another commit - "Add support for TinyALSA"

Yes, please, please @gbarkadiusz move this wrapper to a separate PR. So we can finally close this PR. It was not an easy one and it deserved time but no need to add new stuff. Why add it so late? 

This is a bit ironic: it's common for submitters to complain that reviewers keep "moving the goal post". This time it's the exact way around! :-)

https://github.com/zephyrproject-rtos/zephyr/pull/83839/files

golowanow replied:
> > I think check-tinyalsa.sh should be moved to another commit - "Add support for TinyALSA"
> 
> Yes, please, please @gbarkadiusz move this wrapper to a separate PR. So we can finally close this PR. 

There are two commits currently in this PR, and for `check-tinyalsa.sh` either the first commit ("Add support for TinyALSA") looks better, or just one more commit here for that wrapper script.

IMO, having TinyALSA feature in one PR is better/easier to keep its implementation consistent and tested.

Let's finalize this PR with minimal changes to what's already done here and discussed.

golowanow commented on case-lib/lib.sh:
```diff
@@ -752,17 +752,83 @@ func_lib_check_pa()
+
+# Function to extract the numeric format value from the PCM sample formats
+# There is passes PCM sample format while using ALSA tool (arecord)
+# While using tiny asla (tinycap) we need to convert PCM sample fomrat to bites
+# suporrted by tiny alsa -b parameter.
```
```suggestion
# While using TinyALSA (tinycap -b) we need to convert PCM sample fomrat to bits.
```
please use 'TinyALSA' and 'ALSA' naming consistently in this PR (where it is not a variable name, but some description)

golowanow commented:
>  ... came in the most recent force-push; _after_ I approved.

perhaps this repo/branch has (default?) setting of keep (no dismiss) approvals on force-push ?

marc-hb commented on test-case/check-tinyalsa.sh:
```diff
@@ -0,0 +1,34 @@
+
+# Function to check if a command is available
+check_command() {
+    if command -v "$1" &> /dev/null; then
```
`type "$1"` does it all. I don't understand why some people use `command`.

https://lwn.net/Articles/874049/

marc-hb reviewed (commented):
> perhaps this repo/branch has (default?) setting of keep (no dismiss) approvals on force-push ?

Yes, feel free to change it (if you can find it!)

There is no "good" setting IMHO:

- auto-dismissal increases review overhead, especially when submitters don't "batch" force-pushes but do it immediately after each review.

- Preserving the approval state runs the risk of submitters "sneaking" large changes after approval - intentionally or not!

gbarkadiusz commented:
I accidently added shellcheck to that pr with last push force.
I've applied all above suggestions. @golowanow Please take a look again.

golowanow commented:
@gbarkadiusz you still have `shellcheck` in this PR: added by the first commit 4df594cbea4543f41 and removed in the last commit 5e56a0aba0a

marc-hb commented:
> @gbarkadiusz you should simply upgrade the entire distribution in your WSL container; not just shellcheck.

In the meantime:
- Don't install shellcheck inside your clone
- Do not use "blind" git commands like `git commit -a` or `git add *`

Either of these would have avoided the issue.

golowanow reviewed (commented):
I think with introducing TinyALSA, it needs 

1) check `tinyalsa` and `sox` packages at https://github.com/thesofproject/sof-test/blob/main/env-check.sh

and

2) update https://github.com/thesofproject/sof-test/blob/main/README.md?plain=1#L15

marc-hb commented on case-lib/lib.sh:
```diff
@@ -752,17 +752,82 @@ func_lib_check_pa()
+    if [[ "$SOF_ALSA_TOOL" = "tinyalsa" ]]; then
+        dlogc "tinyplay $*"
+        # shellcheck disable=SC2154
+        sox -n -r "$rate" -c "$channel" noise.wav synth "$duration" white
```
`set -e` is very useful, but it does not work in every situation. There is a number of documented exceptions / situations where it does nothing. Like in C, it's good to explicitly check the exit status of all "important" test commands like this:

```shell
    important_test_command ... || die 'important command ...."
```

More context in:
- https://github.com/thesofproject/sof-test/issues/312

golowanow replied:
better to add an explicit check for `sox` result here

marc-hb commented on case-lib/lib.sh:
```diff
@@ -752,17 +752,82 @@ func_lib_check_pa()
-    # shellcheck disable=SC2086
-    aplay $SOF_ALSA_OPTS $SOF_APLAY_OPTS "$@"
+    if [[ "$SOF_ALSA_TOOL" = "tinyalsa" ]]; then
+        dlogc "tinyplay $*"
```
Here's a logging trick that I use a lot:

```shell
  ( set -x
    sox -n -r "$rate" -c "$channel" noise.wav syn
  )
```

It has some limitations but it works  fine in the majority of cases.

marc-hb commented:
I really don't understand the insistence on including `check-tinyalsa.sh` in this PR. It's a different file so postponing it does not even require any git rebase skillz...

golowanow commented:
> I really don't understand the insistence on including `check-tinyalsa.sh` in this PR. It's a different file so postponing it does not even require any git rebase skillz...

we already use this wrapper script to run TinyALSA tests on CI with this PR

marc-hb commented:
> we already use this wrapper script to run TinyALSA tests on CI with this PR

It's convoluted to prefix with a wrapper script when prefixing with the variable is enough; strange.

This "check-tinyalsa.sh" can be a separate (and smaller) test that runs first = exactly what its name sounds. Similar to the sanity checks found in `test-case/verify-kernel-boot-log.sh`: they are run once at the start, not every time.

golowanow commented:
@gbarkadiusz please let me know when all my comments are addressed to review again, and I hope, it would be the last round in this review.

> > we already use this wrapper script to run TinyALSA tests on CI with this PR
> 
> It's convoluted to prefix with a wrapper script when prefixing with the variable is enough; strange.

@marc-hb, this way, as a wrapper script, it allows for both alsa and tinyalsa test case flavors to coexist in our current CI environment's configuration.

redzynix commented on README.md:
```diff
@@ -52,12 +58,14 @@ Usage: ./check-playback.sh [OPTION]
+  - SOF_APLAY_OPTS and SOF_ARECORD_OPTS contain optional parameters passed additionally on play and record respectively.
+These options are applied to the selected tool (alsa or tinyalsa) based on the value of SOF_ALSA_TOOL 
+
+Warning these environments variables do NOT support parameters
```
```suggestion
Warning, these environment variables do NOT support parameters
```

marc-hb commented:
> @marc-hb, this way, as a wrapper script, it allows for both alsa and tinyalsa test case flavors to coexist in our current CI environment's configuration.

I still don't get it sorry. Anywhere you write:

```
check-tinyalsa.sh sometest.sh test args
```

You can write instead:

```
SOF_ALSA_TOOL=tinyalsa sometest.sh test args
```

... and the effect is the same - but with fewer indirections and complexity.

(The environment variable can also be used in other ways that the wrapper can't do)

marc-hb reviewed (changes requested):
> I still don't get it sorry.

Thanks @golowanow for the private discussion and explanation. Trying to summarize it here:

sof-test has always been invoked by an internal, higher level, Intel test framework that manages test plans and test runs. This framework has some shortcomings, probably the most important (and somewhat relevant here) being sof-framework/issues/602.

That framework does NOT allow environment variables when defining test plans. The interface only lets the user input a test script and script arguments. So, this wrapper script is a way to disguise and shoehorn an environment variable as a test script, sneaking the real test script as a parameter.

Considering this situation, I cannot think of any better to achieve what this wrapper does. BUT, I think the actual purpose of this script should be much clearer. So interactive users and better test frameworks do NOT start using it when they do not need to.

- The wrapper should not be called `check-tinyalsa.sh` which sounds like something good. It should be called something like `tinyalsa-wrapper.sh`.
- The comments at the top of the wrapper should clearly state something like "This is a wrapper that allows to define SOF_ALSA_TOOL in frameworks that cannot define environment variables. Do not use this script if you can just define SOF_ALSA_TOOL.
- The "checking bits" are useful and a very different feature! They should be in a separate "check-tinyalsa.sh` script that performs ONLY the checks and that everyone can use. The former script can call the latter script.

gbarkadiusz commented:
@marc-hb Please take a look.

marc-hb commented on case-lib/lib.sh:
```diff
@@ -752,17 +752,87 @@ func_lib_check_pa()
+# While using TinyALSA (tinycap -b) we need to convert PCM sample fomrat to bits.
+extract_format_number() {
+    # (e.g., extracting '16' from 'S16_LE')
+    format=$(printf '%s' "$1" | grep '[0-9]\+' -o)
```
`format` is a very common name for a global variable.

This function returns a single value so it could just print it instead and avoid a global?

gbarkadiusz replied:
Fixed.

marc-hb commented on case-lib/lib.sh:
```diff
@@ -752,17 +752,87 @@ func_lib_check_pa()
+    dev=$(func_pipeline_parse_value "$idx" dev)
+    pcm=$(func_pipeline_parse_value "$idx" pcm)
+    type=$(func_pipeline_parse_value "$idx" type)
+    snd=$(func_pipeline_parse_value "$idx" snd)
```
That's a lot of global variables with very short names...

gbarkadiusz replied:
I know, but I only moved this section from:
https://github.com/thesofproject/sof-test/pull/1260/files#diff-ac523ac7afa0a9a96eb0ee6c918c10f317d1b201b7124bf10bcf98b611b7324dL78
https://github.com/thesofproject/sof-test/pull/1260/files#diff-4c371b42f9f03814333bc7e33dce34faf542f27087fcf0b5702029f55f630edbL71
to an additional function in lib.sh file.

marc-hb commented on case-lib/lib.sh:
```diff
@@ -752,17 +752,87 @@ func_lib_check_pa()
-    aplay $SOF_ALSA_OPTS $SOF_APLAY_OPTS "$@"
+    if [[ "$SOF_ALSA_TOOL" = "tinyalsa" ]]; then
+        # shellcheck disable=SC2154
+	if ! sox -n -r "$rate" -c "$channel" noise.wav synth "$duration" white; then
```
Indentation? Did you use tabs? I think Github does not like them (I don't either)

gbarkadiusz replied:
True, Fixed.

marc-hb commented on case-lib/lib.sh:
```diff
@@ -752,17 +752,87 @@ func_lib_check_pa()
+        # shellcheck disable=SC2086
+        aplay $SOF_ALSA_OPTS $SOF_APLAY_OPTS "$@"
+    else
+        printf '%s' "Unknown alsa tool: $SOF_ALSA_TOOL "
```
This should be fatal, using `die` or similar.

gbarkadiusz replied:
Done.

marc-hb commented on case-lib/lib.sh:
```diff
@@ -752,17 +752,87 @@ func_lib_check_pa()
+    if [[ "$SOF_ALSA_TOOL" = "tinyalsa" ]]; then
+        # shellcheck disable=SC2154
+	if ! sox -n -r "$rate" -c "$channel" noise.wav synth "$duration" white; then
+            printf '%s' "Error: sox command failed." >&2
```
```suggestion
            printf 'Error: sox command failed.\n' >&2
```
or
```suggestion
            dloge 'Error: sox command failed." 
```

gbarkadiusz replied:
Fixed.

marc-hb commented on case-lib/lib.sh:
```diff
@@ -752,17 +752,87 @@ func_lib_check_pa()
+
+    if [[ "$SOF_ALSA_TOOL" = "tinyalsa" ]]; then
+        # shellcheck disable=SC2154
+        extract_format_number "$fmt_elem"
```
Could you add a comment explaining where $fmt_elem comes from? Shellcheck seems to make a good point here....

More global variables :-(

marc-hb commented on case-lib/lib.sh:
```diff
@@ -752,17 +752,87 @@ func_lib_check_pa()
+        # shellcheck disable=SC2086
+        arecord $SOF_ALSA_OPTS $SOF_ARECORD_OPTS "$@"
+    else
+        printf '%s' "Unknown alsa tool: $SOF_ALSA_TOOL "
```
Should be fatal using `die` or similar.

gbarkadiusz replied:
Fixed.

marc-hb commented on case-lib/lib.sh:
```diff
@@ -1063,15 +1133,31 @@ set_alsa_settings()
+            fi
+        done
+    elif [[ "$SOF_ALSA_TOOL" = "tinyalsa" ]]; then
+        tinymix -D0 controls | sed -e "s/^.*'\(.*\)'.*/\1/" |grep -E 'PGA|gain' |
```
Nit: later, it would be nice to de-duplicate the sed+grep bit with a function. No need to do it now.

marc-hb reviewed (commented):
I keep finding issues sorry. I don't which it is:
- I didn't see them before
- The code has changed
- I reported them before but my comments got "lost"

Nothing major.
